tests/Benchmarks/System.Text.Json/JsonWriterPerf.cs

```diff
@@ -0,0 +1,190 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Text.Formatting;
+
+using static System.Text.Json.Benchmarks.Helper;
+
+namespace System.Text.Json.Benchmarks
+{
+    [MemoryDiagnoser]
+    public class JsonWriterPerf
+    {
+        private const int ExtraArraySize = 500;
+
+        private const int BufferSize = 1024 + (ExtraArraySize * 64);
+
+        private ArrayFormatter _arrayFormatter;
+        private SymbolTable _symbolTable;
+        private MemoryStream _stream;
+        private StreamWriter _writer;
+
+        [Params(EncoderTarget.InvariantUtf8, EncoderTarget.InvariantUtf16)]
+        public EncoderTarget Target;
+
+        [Params(true, false)]
+        public bool Formatted;
+
+        [GlobalSetup]
+        public void Setup()
+        {
+            var enc = Target == EncoderTarget.InvariantUtf8 ? Encoding.UTF8 : Encoding.Unicode;
+            var buffer = new byte[BufferSize];
+            _stream = new MemoryStream(buffer);
+            _writer = new StreamWriter(_stream, enc, BufferSize, true);
+            _symbolTable = GetTargetEncoder(Target);
+            _arrayFormatter = new ArrayFormatter(BufferSize, _symbolTable);
+        }
+
+        [Benchmark]
+        public void WriterSystemTextJsonBasic()
+        {
+            _arrayFormatter.Clear();
+            WriterSystemTextJsonBasic(Formatted, _arrayFormatter);
+        }
+
+        [Benchmark]
+        public void WriterNewtonsoftBasic()
+        {
+            _stream.Seek(0, SeekOrigin.Begin);
+            WriterNewtonsoftBasic(Formatted, _writer);
+        }
+
+        [Benchmark]
+        public void WriterSystemTextJsonHelloWorld()
+        {
+            _arrayFormatter.Clear();
+            WriterSystemTextJsonHelloWorld(Formatted, _arrayFormatter);
+        }
+
+        [Benchmark]
+        public void WriterNewtonsoftHelloWorld()
+        {
+            _stream.Seek(0, SeekOrigin.Begin);
+            WriterNewtonsoftHelloWorld(Formatted, _writer);
+        }
+
+        //[Benchmark]
+        [ArgumentsSource(nameof(GetWriterSystemTextJsonParameters))]
+        public void WriterSlowSystemTextJsonHelloWorld(bool formatted, ArrayFormatter formatter)
+        {
+            formatter.Clear();
+            WriterSystemTextJsonHelloWorld(formatted, formatter);
+        }
+
+        //[Benchmark]
+        [ArgumentsSource(nameof(GetWriterSystemTextJsonParameters))]
+        public void WriterSlowSystemTextJsonBasic(bool formatted, ArrayFormatter formatter)
+        {
+            formatter.Clear();
+            WriterSystemTextJsonBasic(formatted, formatter);
+        }
+
+        // Cannot use this since ArrayFormatter is not a compile time constant
+        // System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.MissingMethodException: No parameterless constructor defined for this object.
+        public IEnumerable<object[]> GetWriterSystemTextJsonParameters()
```

Comment:
@adamsitnik, how can I create arguments that are not compile-time constants? I tried to implement IParam instead but was running into some issues. Any suggestions to get me on the right track?

Reply:
@ahsonkhan I must admit that this is one of the BDN pitfalls (the compilation time requirement for arguments). What you did (using primitive field params) is the best way to do it as of today. Implementing `IParam` requires more work.

I will try to think about more user friendly solution, but as of today I don't have any better idea.